dali/python/nvidia/dali/plugin/base_iterator.py

```diff
@@ -0,0 +1,185 @@
+            shards_beg = np.floor(self._shards_id * self._size_no_pad / self._shards_num).astype(np.int)
+            shards_end = np.floor((self._shards_id + 1) * self._size_no_pad / self._shards_num).astype(np.int)
+            left = self.batch_size - (self._counter - (shards_end - shards_beg))
+            if_drop = np.less(left, self.batch_size)
+        return if_drop, left
+
+    def reset(self):
+        """
+        Resets the iterator after the full epoch.
+        DALI iterators do not support resetting before the end of the epoch
+        and will ignore such request.
+        """
+        if self._counter >= self._size or self._size < 0:
+            if self._fill_last_batch and not self._last_batch_padded:
+                if self._reader_name:
+                    # accurate way
+                    # get the number of samples read in this epoch by each GPU
+                    # self._counter had initial value of min(self._counter_per_gpu) so substract this to get the actual value
+                    self._counter -= min(self._counter_per_gpu)
+                    self._counter_per_gpu = self._counter_per_gpu + self._counter
+                    # check how much each GPU read ahead from next shard, as shards have different size each epoch
+                    # GPU may read ahead or not
+                    self._counter_per_gpu = self._counter_per_gpu - self._shard_sizes
+                    # to make sure that in the next epoch we read the whole shard we need to set start value to the smallest one
+                    self._counter = min(self._counter_per_gpu)
+                else:
+                    # legacy way
+                    self._counter = self._counter % self._size
+            else:
+                self._counter = 0
+            # advance to the next shard
+            if self._reader_name:
+                if not self._is_stick_to_shard:
+                    # move shards id for wrapped pipeliens
+                    self._shards_id = (self._shards_id + 1) % self._shards_num
+                # revaluate _size
+                if self._fill_last_batch and not self._last_batch_padded:
+                    # move all shards ids GPU ahead
+                    if not self._is_stick_to_shard:
+                        self._shard_sizes = np.roll(self._shard_sizes, 1)
+                    # check how many samples we need to reach from each shard in next epoch per each GPU
+                    # taking into account already read
+                    read_in_next_epoch = self._shard_sizes - self._counter_per_gpu
+                    # get the maximmum number of samples and round it up to full batch sizes
+                    self._size = math.ceil(max(read_in_next_epoch) / self.batch_size) * self.batch_size
+                    # in case some epoch is skipped because we have read ahead in this epoch so much
+                    # that in the next one we done already
+                    if self._size == 0:
```

Comment:
Shouldn't we just skip to the next epoch? (I don't really know).

Reply:
I added code to properly move to epoch one after the next, empty one. I don't think we should allow iterator to have epoch of size 0 - this can mess up with the user code that would calculate very short epoch an mess up with some loss or something calculation.

Reply:
That seems like better approach, but wouldn't we still converge to 0-sized epoch? I mean, to get here, we need to read more and more in advance, and if the size of the dataset is not divisable by the batch and the remainder is also some weird number (such that we will always have to read something ahead) we will have shorter and shorter epochs. On grand scale of things this will probably be really slow progress, but with some toy examples you can probably get some weird results.